// Author: Alexander Thomson (thomson@cs.yale.edu)
// Author: Kun Ren (kun.ren@yale.edu)
//
// TODO(alex): Fix types for read_set and write_set.

message TxnProto {
  // Globally unique transaction id, specifying global order.
  required int64 txn_id = 1;
  optional int64 local_txn_id = 2;
  required int64 involved_nodes = 3;
  required int64 batch_number = 4;
  optional bool uncertain = 5;

  // Specifies which stored procedure to invoke at execution time.
  optional int32 txn_type = 10;

  // Isolation level at which to execute transaction.
  //
  // Note: Currently only full serializability is supported.
  
  // True if transaction is known to span multiple database nodes.
  optional bool multipartition = 12;

  // Keys of objects read (but not modified) by this transaction.
  repeated bytes read_set = 20;

  // Keys of objects modified (but not read) by this transaction.
  repeated bytes write_set = 21;

  // Keys of objects read AND modified by this transaction.
  repeated bytes read_write_set = 22;

  // Arguments to be passed when invoking the stored procedure to execute this
  // transaction. 'arg' is a serialized protocol message. The client and backend
  // application code is assumed to know how to interpret this protocol message
  // based on 'txn_type'.
  optional bytes arg = 23;
  
  required int64 seed = 24;	
  optional int64 start_time = 25;

  // Node ids of nodes that participate as readers and writers in this txn.
  repeated int32 readers = 40;
  repeated int32 writers = 41;
}

